New string with empty metadata value chooses wrong constructor

When trying to create a new string and passing in a metadata value, MSBuild tries to bind the the constructor that takes no arguments.  Unfortunately `String` does not have a parameterless constructor and so you get an error.  MSBuild should be using `string.Empty` for these values but for some reason it goes for the wrong overload.

The workaround is to surround the expression with quotes to ensure that the correct overload is chosen.

The below hack is in place because its the only way to call instance methods on metadata values. 
(Related https://github.com/microsoft/msbuild/issues/1155)

### Steps to reproduce

Either include a project sample, attach a zipped project, or provide IDE / CLI steps to create the project and repro the behaviour. Example of a project sample:

Project file
```xml
<Project>
  <ItemGroup>
    <None Include="SomeFolder\**"
          Link="$([System.String]::new(%(RecursiveDir)).Replace(`foo`, `bar`))%(Filename)%(Extension)" />
  </ItemGroup>

  <Target Name="Build">
    <Message Text="%(None.Identity)" Importance="High" />
  </Target>
</Project>
```

Directory contents:
```
│   test.proj
│
└───SomeFolder
        test.txt
```

Command line
```
msbuild /bl
```
### Expected behavior
MSBuild creates a string and calls `.Replace()`.

### Actual behavior
```
error MSB4184: The expression "[System.String]::new()" cannot be evaluated. Attempted to access a missing method.
```

### Workaround
You can places quotes around `%(RecursiveDir)` to force MSBuild to use the overload of `new String()` that accepts a string.
```diff
- $([System.String]::new(%(RecursiveDir))
+ $([System.String]::new(`%(RecursiveDir)`)
```
### Environment data
`msbuild /version` output:

OS info:

If applicable, version of the tool that invokes MSBuild (Visual Studio, dotnet CLI, etc):

--- a/src/Build.UnitTests/Evaluation/Expander_Tests.cs
+++ b/src/Build.UnitTests/Evaluation/Expander_Tests.cs
@@ -2489,7 +2489,7 @@ namespace Microsoft.Build.UnitTests.Evaluation
         /// Expand property function that is defined (on CoreFX) in an assembly named after its full namespace.
         /// </summary>
         [Fact]
-        public void PropertyStaticFunctioLocatedFromAssemblyWithNamespaceName()
+        public void PropertyStaticFunctionLocatedFromAssemblyWithNamespaceName()
         {
             PropertyDictionary<ProjectPropertyInstance> pg = new PropertyDictionary<ProjectPropertyInstance>();
 
@@ -3430,6 +3430,26 @@ namespace Microsoft.Build.UnitTests.Evaluation
             string result = expander.ExpandIntoStringLeaveEscaped(@"$([System.IO.Path]::Combine($(SomePath),%(Compile.Identity)))", ExpanderOptions.ExpandAll, MockElementLocation.Instance);
 
             Assert.Equal(Path.Combine(s_rootPathPrefix, "some", "path", "fOo.Cs"), result);
+        }
+
+        /// <summary>
+        /// Expand a property function which is a string constructor referencing item metadata.
+        /// </summary>
+        /// <remarks>
+        /// Note that referencing a non-existent metadatum results in binding to a parameter-less String constructor. This constructor
+        /// does not exist in BCL but it is special-cased in the expander logic and handled to return an empty string.
+        /// </remarks>
+        [Theory]
+        [InlineData("language", "english")]
+        [InlineData("nonexistent", "")]
+        public void PropertyStringConstructorConsumingItemMetadata(string metadatumName, string metadatumValue)
+        {
+            ProjectHelpers.CreateEmptyProjectInstance();
+            var expander = CreateItemFunctionExpander();
+
+            string result = expander.ExpandIntoStringLeaveEscaped($"$([System.String]::new(%({metadatumName})))", ExpanderOptions.ExpandAll, MockElementLocation.Instance);
+
+            result.ShouldBe(metadatumValue);
         }
 
         /// <summary>

--- a/src/Build/Evaluation/Expander.cs
+++ b/src/Build/Evaluation/Expander.cs
@@ -3341,7 +3341,10 @@ namespace Microsoft.Build.Evaluation
                     // need to locate an appropriate constructor and invoke it
                     if (String.Equals("new", _methodMethodName, StringComparison.OrdinalIgnoreCase))
                     {
-                        functionResult = LateBindExecute(null /* no previous exception */, BindingFlags.Public | BindingFlags.Instance, null /* no instance for a constructor */, args, true /* is constructor */);
+                        if (!TryExecuteWellKnownConstructorNoThrow(out functionResult, args))
+                        {
+                            functionResult = LateBindExecute(null /* no previous exception */, BindingFlags.Public | BindingFlags.Instance, null /* no instance for a constructor */, args, true /* is constructor */);
+                        }
                     }
                     else
                     {
@@ -4040,6 +4043,33 @@ namespace Microsoft.Build.Evaluation
                     LogFunctionCall("PropertyFunctionsRequiringReflection", objectInstance, args);
                 }
 
+                return false;
+            }
+
+            /// <summary>
+            /// Shortcut to avoid calling into binding if we recognize some most common constructors.
+            /// Analogous to TryExecuteWellKnownFunction but guaranteed to not throw.
+            /// </summary>
+            /// <param name="returnVal">The instance as created by the constructor call.</param>
+            /// <param name="args">Arguments.</param>
+            /// <returns>True if the well known constructor call binding was successful.</returns>
+            private bool TryExecuteWellKnownConstructorNoThrow(out object returnVal, object[] args)
+            {
+                returnVal = null;
+
+                if (_receiverType == typeof(string))
+                {
+                    if (args.Length == 0)
+                    {
+                        returnVal = String.Empty;
+                        return true;
+                    }
+                    if (TryGetArg(args, out string arg0))
+                    {
+                        returnVal = arg0;
+                        return true;
+                    }
+                }
                 return false;
             }
 

